### PR TITLE
Normalize CV-CUDA Backend

### PR DESCRIPTION
**Pull request by justincdavis**
### Summary

This PR adds the CV-CUDA backend kernel for the `Normalize` transform.

### How to use

```
import cvcuda
import torchvision.transforms.v2.functional as F

cvc_tensor = cvcuda.Tensor((1, 224, 224, 3), cvcuda.Type.F32, cvcuda.TensorLayout.NHWC)
# Dispatches to F.normalize_cvcuda
normalized_tensor = F.normalize(cvc_tensor, [0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
```

### Run unit tests

```
pytest test/test_transforms_v2.py::TestNormalizeCVCUDA
...
60 passed in 0.59s
```

### CODE REVIEW

File: test/test_transforms_v2.py [hunk 1]
```diff
@@ -5527,6 +5537,11 @@ def test_functional(self, make_input):
             (F.normalize_image, torch.Tensor),
             (F.normalize_image, tv_tensors.Image),
             (F.normalize_video, tv_tensors.Video),
+            pytest.param(
+                F._misc._normalize_cvcuda,
+                _import_cvcuda().Tensor,
```
Comment (AntoineSimoulin) [inline test/test_transforms_v2.py]: @justincdavis it seems that `_import_cvcuda().Tensor` is still raising an error if cvcuda is not installed. Maybe we can just use `cvcuda.Tensor` here and see if this works better?

File: torchvision/transforms/v2/functional/_misc.py [hunk 2]
```diff
@@ -72,6 +79,52 @@ def normalize_video(video: torch.Tensor, mean: list[float], std: list[float], in
...
+        raise ValueError(f"Std should have {channels} elements. Got {len(std)}.")
+
+    # CV-CUDA requires float32 tensors for the mean/std parameters
+    # at small batchs, this is costly relative to normalize operation
+    # if CV-CUDA is known to be a backend, could optimize this
+    # For Normalize class:
+    # by creating tensors at class initialization time
+    # For functional API:
+    # by storing cached tensors in helper function with functools.lru_cache (would it even be worth it?)
+    # Since CV-CUDA is 1) not default backend, 2) only strictly faster at large batch size, ignore
```
Comment (AntoineSimoulin) [inline torchvision/transforms/v2/functional/_misc.py]: Maybe we can remove this comment from here but add it as comment in the PR?

File: torchvision/transforms/v2/functional/_misc.py [hunk 3]
```diff
@@ -72,6 +79,52 @@ def normalize_video(video: torch.Tensor, mean: list[float], std: list[float], in
...
+    mt = torch.as_tensor(mean, dtype=torch.float32).reshape(1, 1, 1, channels).cuda()
+    st = torch.as_tensor(std, dtype=torch.float32).reshape(1, 1, 1, channels).cuda()
+    mean_cv = cvcuda.as_tensor(mt, cvcuda.TensorLayout.NHWC)
+    std_cv = cvcuda.as_tensor(st, cvcuda.TensorLayout.NHWC)
```
Comment (AntoineSimoulin) [inline torchvision/transforms/v2/functional/_misc.py]: Do we have to create a `torch.Tensor` and convert it to a `cvcuda.Tensor` here? Can we directly create a `cvcuda.Tensor` from list of floats?

File: torchvision/transforms/v2/functional/_misc.py [hunk 4]
```diff
@@ -72,6 +79,52 @@ def normalize_video(video: torch.Tensor, mean: list[float], std: list[float], in
...
+
+    return cvcuda.normalize(image, base=mean_cv, scale=std_cv, flags=cvcuda.NormalizeFlags.SCALE_IS_STDDEV)
+
+
+if CVCUDA_AVAILABLE:
+    _normalize_cvcuda_registered = _register_kernel_internal(normalize, _import_cvcuda().Tensor)(_normalize_cvcuda)
```
Comment (AntoineSimoulin) [inline torchvision/transforms/v2/functional/_misc.py]: We should be able to remove `_normalize_cvcuda_registered =` here and keep only `_register_kernel_internal(normalize, _import_cvcuda().Tensor)(_normalize_cvcuda)`.

File: test/test_transforms_v2.py [hunk 5]
```diff
@@ -5579,6 +5604,17 @@ def test_correctness_image(self, mean, std, dtype, fn):
 
         assert_equal(actual, expected)
 
+    @pytest.mark.skipif(not CVCUDA_AVAILABLE, reason="test requires CVCUDA")
+    @pytest.mark.parametrize(("mean", "std"), MEANS_STDS)
+    @pytest.mark.parametrize("dtype", [torch.float32])
```
Comment (AntoineSimoulin) [inline test/test_transforms_v2.py]: Maybe we don't need to parametrize if we are only testing against `torch.float32` format?

**Review by AntoineSimoulin — COMMENTED** [review]
Hey @justincdavis, thanks for submitting the PR, this is looking good:) I left some minor changes. I think we mainly need to make sure the tests are passing when cvcuda is not installed!

File: test/test_transforms_v2.py (hunk 1 above)
Comment (justincdavis) [reply test/test_transforms_v2.py]: Thank you for pointing this out! I replaced the actual `cvcuda.Tensor` type with the string `"cvcuda.Tensor"`, then inside the function we resolve the `cvcuda.Tensor` type if we have the corresponding string. LMK if this looks like a reasonable solution!

File: torchvision/transforms/v2/functional/_misc.py (hunk 3 above)
Comment (justincdavis) [reply torchvision/transforms/v2/functional/_misc.py]: Unfortunately, CV-CUDA does not support the creation of GPU data from CPU data (even for np.ndarray). Posted a discussion item in the overall PR about the one potential way to get around the two host->device transfers.

File: test/test_transforms_v2.py (hunk 5 above)
Comment (justincdavis) [reply test/test_transforms_v2.py]: Good point! The reason I left this parameterized is because it is in the future plans for CV-CUDA to support float16. However, we can simply add the parameterize once that support is there.

**Comment by justincdavis** [comment]
Following up from my comment in the `_normalize_cvcuda` function itself. CV-CUDA requires that the mean and scale tensors be on-device when we call `cvcuda.normalize`. This means that a host->device memcpy must occur twice for each normalize call when using CV-CUDA backend. We could attempt to reduce the impact of this by having a helper function which creates the `tuple[cvcuda.Tensor, cvcuda.Tensor]` from the mean/std parameters. Based on what I see in the codebase, this seems like it would be a new feature present in torchvision for a functional transform. 

>     # CV-CUDA requires float32 tensors for the mean/std parameters
>     # at small batchs, this is costly relative to normalize operation
>     # if CV-CUDA is known to be a backend, could optimize this
>     # For Normalize class:
>     # by creating tensors at class initialization time
>     # For functional API:
>     # by storing cached tensors in helper function with functools.lru_cache (would it even be worth it?)
>     # Since CV-CUDA is 1) not default backend, 2) only strictly faster at large batch size, ignore

File: torchvision/transforms/v2/functional/_misc.py (hunk 3 above)
Comment (AntoineSimoulin) [reply torchvision/transforms/v2/functional/_misc.py]: Gotcha, thanks for the details. I think we can keep it as is for now then. We can consider later streamlining this part.

**Comment by AntoineSimoulin** [comment]
Hey @justincdavis, looking good to me. I don't think the failing test is related to this PR. Seems like a false positive alert to me! Can you sign our Contributor License Agreement (c.f. [meta-cla](https://github.com/apps/meta-cla) bot comment in the discussion)?

File: test/test_transforms_v2.py [hunk 6]
```diff
@@ -5552,24 +5552,34 @@ def test_kernel_image(self, mean, std, device):
 
     @pytest.mark.parametrize("device", cpu_and_cuda())
     def test_kernel_image_inplace(self, device):
-        input = make_image_tensor(dtype=torch.float32, device=device)
-        input_version = input._version
+        inpt = make_image_tensor(dtype=torch.float32, device=device)
```
Comment (zy1git) [inline test/test_transforms_v2.py]: Revert as we discussed in [https://github.com/pytorch/vision/pull/9278#discussion_r2590050883](https://github.com/pytorch/vision/pull/9278#discussion_r2590050883)

File: test/test_transforms_v2.py [hunk 7]
```diff
@@ -5552,24 +5552,34 @@ def test_kernel_image(self, mean, std, device):
...
 
-    @pytest.mark.parametrize("make_input", [make_image_tensor, make_image, make_video])
+    @pytest.mark.parametrize(
+        "make_input",
+        [
+            make_image_tensor,
+            make_image,
+            make_video,
+            pytest.param(
+                make_image_cvcuda, marks=pytest.mark.skipif(not CVCUDA_AVAILABLE, reason="test requires CVCUDA")
```
Comment (zy1git) [inline test/test_transforms_v2.py]: See this PR: [https://github.com/pytorch/vision/pull/9305/changes](https://github.com/pytorch/vision/pull/9305/changes)

File: test/test_transforms_v2.py [hunk 8]
```diff
@@ -5579,9 +5589,16 @@ def test_functional(self, make_input):
             (F.normalize_image, torch.Tensor),
             (F.normalize_image, tv_tensors.Image),
             (F.normalize_video, tv_tensors.Video),
+            pytest.param(
+                F._misc._normalize_image_cvcuda,
+                None,
+                marks=pytest.mark.skipif(not CVCUDA_AVAILABLE, reason="test requires CVCUDA"),
```
Comment (zy1git) [inline test/test_transforms_v2.py]: Same as [https://github.com/pytorch/vision/pull/9279/changes#r2613353258](https://github.com/pytorch/vision/pull/9279/changes#r2613353258)

File: test/test_transforms_v2.py [hunk 9]
```diff
@@ -5595,9 +5612,9 @@ def test_functional_error(self):
             with pytest.raises(ValueError, match="std evaluated to zero, leading to division by zero"):
                 F.normalize_image(make_image(dtype=torch.float32), mean=self.MEAN, std=std)
 
-    def _sample_input_adapter(self, transform, input, device):
+    def _sample_input_adapter(self, transform, inpt, device):
         adapted_input = {}
-        for key, value in input.items():
+        for key, value in inpt.items():
```
Comment (zy1git) [inline test/test_transforms_v2.py]: revert inpt

File: test/test_transforms_v2.py [hunk 10]
```diff
@@ -5607,7 +5624,17 @@ def _sample_input_adapter(self, transform, input, device):
...
 
-    @pytest.mark.parametrize("make_input", [make_image_tensor, make_image, make_video])
+    @pytest.mark.parametrize(
+        "make_input",
+        [
+            make_image_tensor,
+            make_image,
+            make_video,
+            pytest.param(
+                make_image_cvcuda, marks=pytest.mark.skipif(not CVCUDA_AVAILABLE, reason="test requires CVCUDA")
```
Comment (zy1git) [inline test/test_transforms_v2.py]: Same as https://github.com/pytorch/vision/pull/9279/changes#r2613353258

File: test/test_transforms_v2.py [hunk 11]
```diff
@@ -5622,14 +5649,33 @@ def _reference_normalize_image(self, image, *, mean, std):
 
     @pytest.mark.parametrize(("mean", "std"), MEANS_STDS)
     @pytest.mark.parametrize("dtype", [torch.float16, torch.float32, torch.float64])
+    @pytest.mark.parametrize(
+        "make_input",
+        [
+            make_image,
+            pytest.param(
+                make_image_cvcuda, marks=pytest.mark.skipif(not CVCUDA_AVAILABLE, reason="test requires CVCUDA")
```
Comment (zy1git) [inline test/test_transforms_v2.py]: Same as https://github.com/pytorch/vision/pull/9279/changes#r2613353258

File: test/test_transforms_v2.py [hunk 12]
```diff
@@ -5622,14 +5649,33 @@ def _reference_normalize_image(self, image, *, mean, std):
...
-        image = make_image(dtype=dtype)
+    def test_correctness_image(self, mean, std, dtype, make_input, fn):
+        if make_input == make_image_cvcuda and dtype != torch.float32:
+            pytest.skip("CVCUDA only supports float32 for normalize")
+
+        image = make_input(dtype=dtype)
 
         actual = fn(image, mean=mean, std=std)
+
+        if make_input == make_image_cvcuda:
```
Comment (zy1git) [inline test/test_transforms_v2.py]: make_input == make_image_cvcuda is checked three times. Maybe we can use a boolean variable for that?

Like `is_cvcuda = make_input == make_image_cvcuda`

File: torchvision/transforms/v2/functional/_misc.py [hunk 13]
```diff
@@ -72,6 +79,44 @@ def normalize_video(video: torch.Tensor, mean: list[float], std: list[float], in
...
 
 
+def _normalize_image_cvcuda(
+    image: "cvcuda.Tensor",
+    mean: list[float],
+    std: list[float],
+    inplace: bool = False,
+) -> "cvcuda.Tensor":
+    cvcuda = _import_cvcuda()
+    if inplace:
```
Comment (zy1git) [inline torchvision/transforms/v2/functional/_misc.py:87]: @NicolasHug Need to check this logic more carefully.

**Review by zy1git — COMMENTED** [review]
I left some comments. Some of them are referred to the merged PRs in the past three weeks.

File: torchvision/transforms/v2/functional/_misc.py [hunk 14]
```diff
@@ -72,6 +77,44 @@ def normalize_video(video: torch.Tensor, mean: list[float], std: list[float], in
...
+        raise ValueError("Inplace normalization is not supported for CVCUDA.")
+
+    # CV-CUDA supports signed int and float tensors
+    # torchvision only supports uint and float, right now CV-CUDA doesnt expose float16, so only check 32
+    # in the future add float16 once exposed in CV-CUDA
+    if not (image.dtype == cvcuda.Type.F32):
+        raise ValueError(f"Input tensor should be a float tensor. Got {image.dtype}.")
+
+    channels = image.shape[3]
```
Comment (NicolasHug) [inline torchvision/transforms/v2/functional/_misc.py]: Nit: we prefer using the `num_` prefix for counted stuff, so this could be `num_channels`.

File: test/test_transforms_v2.py [hunk 15]
```diff
@@ -5614,14 +5637,33 @@ def _reference_normalize_image(self, image, *, mean, std):
...
+        if is_cvcuda:
+            image = F.cvcuda_to_tensor(image)[0].cpu()
+
         expected = self._reference_normalize_image(image, mean=mean, std=std)
 
-        assert_equal(actual, expected)
+        if is_cvcuda:
+            assert_close(actual, expected, rtol=0, atol=1e-6)
+        else:
+            assert_equal(actual, expected)
```
Comment (NicolasHug) [inline test/test_transforms_v2.py:5671]: I'm surprised atol=1e-6 is needed, I thought it was the default for float32. Let's try without it and see if the CI is happy?

File: test/test_transforms_v2.py [hunk 16]
```diff
@@ -5614,14 +5637,33 @@ def _reference_normalize_image(self, image, *, mean, std):
...
+            pytest.skip("CVCUDA only supports float32 for normalize")
+
+        image = make_input(dtype=dtype)
 
         actual = fn(image, mean=mean, std=std)
+
+        if is_cvcuda:
+            image = F.cvcuda_to_tensor(image)[0].cpu()
+
         expected = self._reference_normalize_image(image, mean=mean, std=std)
```
Comment (NicolasHug) [inline test/test_transforms_v2.py:5666]: Note for self: double-checking that this is doing the right conversion. `image` is not a tensor and we're using  `_reference_normalize_image` as the ref, we'll be comparing our cvcuda-tensor to this reference tensor. Seems OK.

File: test/test_transforms_v2.py [hunk 17]
```diff
@@ -5614,14 +5637,33 @@ def _reference_normalize_image(self, image, *, mean, std):
...
+        ],
+    )
     @pytest.mark.parametrize("fn", [F.normalize, transform_cls_to_functional(transforms.Normalize)])
-    def test_correctness_image(self, mean, std, dtype, fn):
-        image = make_image(dtype=dtype)
+    def test_correctness_image(self, mean, std, dtype, make_input, fn):
+        is_cvcuda = make_input is make_image_cvcuda
+
+        if is_cvcuda and dtype != torch.float32:
+            pytest.skip("CVCUDA only supports float32 for normalize")
```
Comment (NicolasHug) [inline test/test_transforms_v2.py]: Instead of a skip, this could be an xfail instead. See https://docs.pytest.org/en/stable/how-to/skipping.html

File: test/test_transforms_v2.py
Comment (NicolasHug) [inline test/test_transforms_v2.py:1]: Note for self: need to check that all the relevant tests have been properly parametrized.

File: torchvision/transforms/v2/functional/_misc.py [hunk 18]
```diff
@@ -72,6 +77,44 @@ def normalize_video(video: torch.Tensor, mean: list[float], std: list[float], in
...
+) -> "cvcuda.Tensor":
+    cvcuda = _import_cvcuda()
+    if inplace:
+        raise ValueError("Inplace normalization is not supported for CVCUDA.")
+
+    # CV-CUDA supports signed int and float tensors
+    # torchvision only supports uint and float, right now CV-CUDA doesnt expose float16, so only check 32
+    # in the future add float16 once exposed in CV-CUDA
+    if not (image.dtype == cvcuda.Type.F32):
+        raise ValueError(f"Input tensor should be a float tensor. Got {image.dtype}.")
```
Comment (NicolasHug) [inline torchvision/transforms/v2/functional/_misc.py:94]: We should have a test that asserts non-float32 leads to an error and check the error message

**Review by NicolasHug — COMMENTED** [review]
Thanks a lot for the PR @justincdavis , I left a review for @zy1git to address.

File: test/test_transforms_v2.py (hunk 15 above)
Comment (zy1git) [reply test/test_transforms_v2.py:5671]: I remove the `rtol = 0, atol=1e-6` and the test passed. However, the default for `assert_close` is `rtol = 1.3e-6, atol=1e-5`, thus if the original implementation `rtol=0, atol=1e-6` can pass the test, the default version definitely can pass. 
I think `atol = 1e-6` is more restricted threshold. I will change it to default in a new commit. Please let me know if you think we need to use the original implementation.

File: torchvision/transforms/v2/functional/_misc.py (hunk 18 above)
Comment (zy1git) [reply torchvision/transforms/v2/functional/_misc.py:94]: Added one test for that: https://github.com/pytorch/vision/pull/9279/changes#diff-9c2dde92db86c123fee225e39b7c1ef96e08a3e79a9dcc9a2d68b21ed51a81d0R5606

File: test/test_transforms_v2.py (hunk 17 above)
Comment (zy1git) [reply test/test_transforms_v2.py]: I changed `pytest.skip` to `pytest.xfail`

**Review by zy1git — COMMENTED** [review]
I addressed comments by pushing a new commit. Feel free to take a look.

File: test/test_transforms_v2.py (hunk 15 above)
Comment (NicolasHug) [reply test/test_transforms_v2.py:5671]: Thank you! It's usually best to rely on the defaults, unless we have a strong reason not-to. Here, the default threshold is strict enough, so we can just use it.

**Review by NicolasHug — APPROVED** [review]
Thanks a ton @justincdavis for the PR, @AntoineSimoulin for the review and @zy1git for finalizing it!
